Docs: Add hedgedoc flavored markdown reference (#1066)
hedgedoc/hedgedoc: docs/content/references/hfm.md

# HedgeDoc Flavored Markdown

> HedgeDoc mostly follows the [CommonMark][commonmark] standard. It shares some extensions with the [GFM][gfm] standard, but for historical reasons HedgeDoc does support not exactly GFM, but a bit more in places (and a bit less in other places).

Reviewer: HedgeDoc has it's own markdown dialect which supports many features from [CommonMark][commonmark] and [Github Flavored Markdown][gfm]. It also adds some new extensions and is missing some.

Author: fixed

> These tables will tell you what exactly we support in HedgeDoc 1.x (HFM 1) and HedgeDoc 2 (HFM 2).

Reviewer: ```suggestion
These tables tell you what exactly we support in HedgeDoc 1.x (HFM 1) and HedgeDoc 2 (HFM 2).
```

Author: done

| code block with language[^highlight]       | ☑️     | ☑️     | ☑️          | ☑️   |
| code block with extra features[^extra-code] | ☑️     | ☑️     |            |     |
| block quote                    | ☑️     | ☑️     | ☑️          | ☑️   |
| name tag (`[name=...]`)        | (☑️)[^in-bq-list]     | ☑️     |            |     |
| time tag (`[time=...]`)        | (☑️)[^in-bq-list]     | ☑️     |            |     |
| color tag (`[color=...]`)      | (☑️)[^in-bq-list]     | ☑️     |            |     |
| unorderd list                  | ☑️     | ☑️     | ☑️          | ☑️   |
| ordered list                   | ☑️     | ☑️     | ☑️          | ☑️   |
| task list                      | ☑️     | ☑️     |            | ☑️   |
| defition list                  | ☑️     | ☑️     |            |     |
| emoji             | [Unicode 6.1][unicode-6] | [Unicode 13][unicode-13] |                      |                      |
| [ForkAwesome][fa] | ☑️ with `<i class='fa'>`  | ☑️ with shortcodes        |                      |                      |
| LaTeX             | ☑️[^mj]                   | ☑️[^kt]                   |                      |                      |

> [^highlight]: Code-blocks with a given language are rendered with syntax-highlighting for the code.

Reviewer: ```suggestion
[^highlight]: Code blocks with a given language are rendered using syntax highlighting.
```

Author: done

[^extra-code]: Several special "language" keywords can be used for rendering diagrams, charts, etc.

> [^in-bq-list]: Use of these tags is only supported within blockquotes or (un)ordered lists.

Reviewer: ```suggestion
[^in-bq-list]: Use of these tags is only supported within blockquotes and (un)ordered lists.
```

Author: done

| footnotes             | ☑️     | ☑️     |                  |                  |
| image                 | ☑️     | ☑️     | ☑️                | ☑️                |
| image with given size | ☑️     | ☑️     | (☑️ with `<img>`) | (☑️ with `<img>`) |
| table of contents     | ☑️     | ☑️     |                  |                  |

## Structural elements

| Feature           | HFM 1                    | HFM 2                    | CommonMark           | GFM                  |
|-------------------|:------------------------:|:------------------------:|:--------------------:|:--------------------:|
| table             | ☑️                        | ☑️                        | (☑️ with `<table>`)   | ☑️                    |
| horizontal line   | ☑️                        | ☑️                        | ☑️                    | ☑️                    |
| collapsable block | (☑️ with `<details>`)     | ☑️                        | (☑️ with `<details>`) | (☑️ with `<details>`) |
| Alerts            | ☑️                        | ☑️                        |                      |                      |

## Embeddings

> HFM1 included support for certain embeddings of external content. These were defined in markdown by the uncommon `{%keyword parameter %}` syntax.
> 
> Instead of this uncommon syntax, HFM2 uses just plain links to external content and creates embeddings for supported providers.

Reviewer: ```suggestion
HFM 1 includes support for certain embeddings of external content by using the `{%keyword parameter %}` syntax. To increase the readability of the markdown code we decided that HFM 2 should just use plain links if possible.
```

Author: done

| Feature                  | HFM 1 | HFM 2              | CommonMark | GFM |
|--------------------------|:-----:|:------------------:|:----------:|:---:|
| PDF (`{%pdf ... %}`)     | ☑️     | removed            |            |     |
| [YouTube][youtube] (`{%youtube ... %}`)       | ☑️     | with plain link[^embed] |            |     |
| [Vimeo][vimeo] (`{%vimeo ... %}`)          | ☑️     | with plain link[^embed] |            |     |
| [Slideshare][slideshare] (`{%slideshare ... %}`) | ☑️     | removed            |            |     |
| [Speakerdeck][speakerdeck] (`{%speakerdeck ... %}`) | ☑️     | removed            |            |     |
| [GitHub Gist][gist] (`{%gist ... %}`)             | ☑️     | with plain link[^embed] |            |     |

> [^embed]: The special syntax is deprecated but will continue to work. However a plain link to the content will generate the same embedding and is preferred.

Reviewer: ```suggestion
[^embed]: The special syntax from HFM 1 is deprecated but will still work in HFM 2. However a plain link to the content should be preferred.
```

Author: I incorporated this, but changed the last words to `should be used.` instead of `should be preferred`.

Reviewer 2: How about making it must be preferred? This would make it clear that these might drop out of support any time in future.

Author: I think I like the current wording best
